compiler/src/lnntoamm/Types.ts

```diff
@@ -123,127 +193,279 @@ export default abstract class Type implements Equalable {
   }
 }
 
-class Builtin extends Type {
+class Opaque extends Type implements Generalizable {
+  // if any values are `null` that means that this isn't an instantiable type
+  // and should be treated like the type needs to be duped
+  generics: GenericArgs;
+
   get ammName(): string {
-    return this.name;
+    let generics = '';
+    if (Object.keys(this.generics).length !== 0) {
+      const genNames = new Array<string>();
+      for (const [tyVar, ty] of Object.entries(this.generics)) {
+        if (ty === null) {
+          genNames.push(tyVar);
+        } else {
+          genNames.push(ty.ammName);
+        }
+      }
+      generics = '<' + genNames.join(', ') + '>';
+    }
+    return this.name + generics;
   }
 
-  constructor(name: string) {
+  constructor(name: string, generics: string[]) {
     super(name);
+    this.generics = {};
+    generics.forEach((g) => (this.generics[g] = null));
   }
 
-  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
-    if (ty instanceof Builtin) {
-      return this.eq(ty);
-    } else if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this, scope);
-    } else if (ty instanceof HasOperator) {
-      return Has.operator(ty, scope, this).length !== 0;
-    } else if (ty instanceof HasMethod) {
-      return Has.method(ty, scope, this).length !== 0;
-    } else if (ty instanceof HasField) {
-      return TODO('add field support for builtins');
-    } else if (ty instanceof Interface) {
-      // FIXME: once builtins can have fields
+  compatibleWithConstraint(that: Type, scope: Scope): boolean {
+    if (this.eq(that)) {
+      return true;
+    }
+    if (that instanceof Opaque) {
+      const thisGens = Object.values(this.generics);
+      const thatGens = Object.values(that.generics);
+      if (this.name !== that.name || thisGens.length !== thatGens.length) {
+        return false;
+      }
       return (
-        ty.fields.length === 0 &&
-        ty.methods.every((m) => this.compatibleWithConstraint(m, scope)) &&
-        ty.operators.every((o) => this.compatibleWithConstraint(o, scope))
+        this.name === that.name &&
+        thisGens.length === thatGens.length &&
+        thisGens.every((thisGen, ii) => {
+          const thatGen = thatGens[ii];
+          if (thisGen === null || thatGen === null) {
+            return true;
+          } else {
+            return thisGen.compatibleWithConstraint(thatGen, scope);
+          }
+        })
       );
+    } else if (that instanceof Interface || that instanceof OneOf) {
+      return that.compatibleWithConstraint(this, scope);
+    } else if (that instanceof HasField) {
+      return false;
```

Comment:
Is this one of the ones that should be labeled "TODO: Requires struct types"?

Reply:
Opaque types by definition don't have any accessible fields - built-in types that expose fields should still use the `Struct` type